### PR TITLE
LLVM bump: move almost entirely to std::optional, fixes

### PR DESCRIPTION
**Pull request by dtzSiFive**
Use of "Optional" in tablegen files is still preserved (for now) as it is in upstream MLIR.

clang-tidy likely will be unhappy for existing problems (e.g., recursion), but the other checks were addressed.

### CODE REVIEW

**Comment by dtzSiFive** [comment]
I'll pull out the fixes/touchups shortly.

**Comment by dtzSiFive** [comment]
The "uploadBinaries" workflow successfully ran on this: https://github.com/llvm/circt/actions/runs/3761021813 .

(it fails on main because of a minor upstream issue in a test that I fixed yesterday).

File: include/circt/Dialect/HW/HWPasses.h [hunk 1]
```diff
@@ -15,8 +15,8 @@
 
 #include "mlir/Pass/Pass.h"
 #include "mlir/Pass/PassRegistry.h"
-#include "llvm/ADT/Optional.h"
 #include <memory>
+#include <optional>
```
Comment (seldridge) [inline include/circt/Dialect/HW/HWPasses.h:19]: This seems like it doesn't belong here and should be defined in the passes that need it. Out-of-scope for this PR.

File: lib/Dialect/FIRRTL/FIRRTLTypes.cpp [hunk 2]
```diff
@@ -1126,38 +1128,32 @@ void FIRRTLDialect::registerTypes() {
...
-        .Case<IntType>([&](IntType iType) {
-          auto retval = iType.getWidth();
-          if (retval)
-            return llvm::Optional<int64_t>(*retval);
-          else
-            return llvm::Optional<int64_t>(std::nullopt);
+            return std::nullopt;
+          return *w * vector.getNumElements();
         })
+        .Case<IntType>([&](IntType iType) { return iType.getWidth(); })
```
Comment (seldridge) [inline lib/Dialect/FIRRTL/FIRRTLTypes.cpp:1154]: Nice that this simplifies.

**Review by seldridge — APPROVED** [review]
Nice!

File: lib/Dialect/FIRRTL/FIRRTLTypes.cpp (hunk 2 above)
Comment (dtzSiFive) [reply lib/Dialect/FIRRTL/FIRRTLTypes.cpp:1154]: Yeah! It does perform an extension from the int32_t optional in `getWidth()` which didn't work before with `llvm::Optional`? (`Optional<int32_t>` -> `Optional<int64_t>`).

**Comment by dtzSiFive** [comment]
Thanks for looking through all the changes, even if mostly mechanical :+1:.

**Pull request merged**